File: Back-End/backend/src/models/Holding.ts
import { db } from '../global/firebase';
import { Timestamp, FieldValue } from 'firebase-admin/firestore';

export interface HoldingInput {
  stockId: string;
  stockName?: string;
  sharesHeld: number;
  avgCost: number;
  totalCost: number;
  realizedProfit: number;
  costMethod: string;
}

export class Holding {
  stockId!: string;
  stockName?: string;       // 持久化於 Firestore，recalculate 時寫入
  sharesHeld!: number;
  avgCost!: number;
  totalCost!: number;
  realizedProfit!: number;
  costMethod!: string;
  updatedAt!: string;
  sortIndex!: number;

  // 即時資料由 Controller 注入，不存 Firestore
  currentPrice?: number;
  change?: number;
  changePercent?: number;

  private static readonly col = db.collection('holdings');

  static async findAll(): Promise<Holding[]> {
    const snap = await this.col.get();
    const holdings = snap.docs.map(doc => Holding.fromSnapshot(doc));
    return holdings.sort((a, b) => a.sortIndex - b.sortIndex);
  }

  static async reorder(order: string[]): Promise<void> {
    const batch = db.batch();
    order.forEach((stockId, index) => {
      batch.update(this.col.doc(stockId), { sort_index: index });
    });
    await batch.commit();
  }

  static async findById(stockId: string): Promise<Holding | null> {
    const doc = await this.col.doc(stockId).get();
    return doc.exists ? Holding.fromSnapshot(doc) : null;
  }

  static async upsert(input: HoldingInput): Promise<void> {
    const payload: Record<string, unknown> = {
      stock_id:        input.stockId,
      shares_held:     input.sharesHeld,
      avg_cost:        input.avgCost,
      total_cost:      input.totalCost,
      realized_profit: input.realizedProfit,
      cost_method:     input.costMethod,
      updated_at:      FieldValue.serverTimestamp(),
    };
    if (input.stockName) payload.stock_name = input.stockName;
    await this.col.doc(input.stockId).set(payload, { merge: true });
  }

  static async batchUpsert(inputs: HoldingInput[]): Promise<void> {
    const batch = db.batch();
    for (const input of inputs) {
      const ref = this.col.doc(input.stockId);
      const payload: Record<string, unknown> = {
        stock_id:        input.stockId,
        shares_held:     input.sharesHeld,
        avg_cost:        input.avgCost,
        total_cost:      input.totalCost,
        realized_profit: input.realizedProfit,
        cost_method:     input.costMethod,
        updated_at:      FieldValue.serverTimestamp(),
      };
      if (input.stockName) payload.stock_name = input.stockName;
      batch.set(ref, payload, { merge: true });
    }
    await batch.commit();
  }

  private static fromSnapshot(doc: FirebaseFirestore.DocumentSnapshot): Holding {
    const d = doc.data()!;
    const h = new Holding();
    h.stockId        = doc.id;
    h.stockName      = d['stock_name'] || undefined;
    h.sharesHeld     = d['shares_held'];
    h.avgCost        = d['avg_cost'];
    h.totalCost      = d['total_cost'];
    h.realizedProfit = d['realized_profit'];
    h.costMethod     = d['cost_method'] ?? 'preserve_method';
    const ua         = d['updated_at'];
    h.updatedAt      = ua instanceof Timestamp ? ua.toDate().toISOString() : new Date().toISOString();
    h.sortIndex      = d['sort_index'] ?? 0;
    return h;
  }
}

File: Back-End/shioaji-service/models/holding.py
from __future__ import annotations
import asyncio
from typing import Optional
from google.cloud.firestore import SERVER_TIMESTAMP
from lib.firebase import get_db

COL = "holdings"


def _ts_to_iso(ts) -> Optional[str]:
    return ts.isoformat() if hasattr(ts, "isoformat") else None


def _from_doc(doc) -> dict:
    d = doc.to_dict()
    return {
        "stockId": doc.id,
        "sharesHeld": d.get("shares_held", 0),
        "avgCost": d.get("avg_cost", 0),
        "totalCost": d.get("total_cost", 0),
        "realizedProfit": d.get("realized_profit", 0),
        "costMethod": d.get("cost_method", "preserve_method"),
        "updatedAt": _ts_to_iso(d.get("updated_at")),
        "sortIndex": d.get("sort_index", 0),
    }


async def find_all() -> list[dict]:
    def _run():
        snap = get_db().collection(COL).get()
        docs = [_from_doc(d) for d in snap]
        return sorted(docs, key=lambda h: h["sortIndex"])
    return await asyncio.to_thread(_run)


async def find_by_id(stock_id: str) -> Optional[dict]:
    def _run():
        doc = get_db().collection(COL).document(stock_id).get()
        return _from_doc(doc) if doc.exists else None
    return await asyncio.to_thread(_run)


async def reorder(order: list[str]) -> None:
    def _run():
        db = get_db()
        batch = db.batch()
        for index, stock_id in enumerate(order):
            batch.update(db.collection(COL).document(stock_id), {"sort_index": index})
        batch.commit()
    await asyncio.to_thread(_run)


async def batch_upsert(holdings: list[dict]) -> None:
    def _run():
        db = get_db()
        batch = db.batch()
        for h in holdings:
            ref = db.collection(COL).document(h["stockId"])
            batch.set(ref, {
                "stock_id": h["stockId"],
                "shares_held": h["sharesHeld"],
                "avg_cost": h["avgCost"],
                "total_cost": h["totalCost"],
                "realized_profit": h["realizedProfit"],
                "cost_method": h["costMethod"],
                "updated_at": SERVER_TIMESTAMP,
            }, merge=True)
        batch.commit()
    await asyncio.to_thread(_run)
